dcrpool.go

```diff
@@ -195,27 +176,57 @@ func realMain() error {
 			err := http.ListenAndServe(listenAddr, nil)
 			if err != nil {
 				mpLog.Criticalf(err.Error())
-				p.cancel()
+				cancel()
 			}
 		}()
 	}
 
-	mpLog.Infof("%s version %s (Go version %s %s/%s)", appName,
-		Version, runtime.Version(), runtime.GOOS, runtime.GOARCH)
-	mpLog.Infof("Home dir: %s", cfg.HomeDir)
-	mpLog.Infof("Started dcrpool")
-
 	go func() {
 		select {
-		case <-p.ctx.Done():
+		case <-ctx.Done():
 			return
 
 		case <-interrupt:
-			p.cancel()
+			cancel()
 		}
 	}()
-	p.gui.Run(p.ctx)
-	p.hub.Run(p.ctx)
+
+	// Create a hub and GUI instance.
+	hub, err := newHub(cfg, db, cancel)
+	if err != nil {
+		mpLog.Errorf("unable to initialize hub: %v", err)
+		return err
+	}
+	gui, err := newGUI(cfg, hub)
+	if err != nil {
+		mpLog.Errorf("unable to initialize GUI: %v", err)
+		return err
+	}
+
+	// Run the GUI in the background.
+	go gui.Run(ctx)
+
+	// Run the hub.  This will block until the context is cancelled.
+	runHub := func(ctx context.Context, h *pool.Hub) error {
+		// Ideally these would go into hub.Run, but the tests don't work
+		// properly with this code there due to their tight coupling.
+		if err := h.Connect(ctx); err != nil {
+			return fmt.Errorf("unable to establish node connections: %w", err)
+		}
+
+		if err := h.FetchWork(ctx); err != nil {
+			return fmt.Errorf("unable to get work from consensus daemon: %w", err)
+		}
+
+		h.Run(ctx)
+		return nil
+	}
+	if err := runHub(ctx, hub); err != nil {
+		// Ensure the GUI is signaled to shutdown.
```

Comment:
Despite calling cancel here, the gui has no waitgroup (or any other mechanism of signalling that it has completed shutdown) so there is no guarantee that it will shutdown completely before the process terminates. Thats an existing problem though, and we can fix that in a later PR.

Reply:
Right.  My hope is that separating it out like this helps show the problem with the current design where it passes the cancel func down deep into the innards of something that does not control the overall process and thus shouldn't be able to just pull the rug out from everything as it's doing now.